semantic-python/src/Language/Python/Core.hs

```diff
@@ -108,20 +123,74 @@ deriving via CompileSum (Either l r) instance (Compile l, Compile r) => Compile
+-- c will be the terminal node. It is never an assignment.
+type RHS a = Either (Py.Assignment a) (Either (Py.AugmentedAssignment a) (Desugared a))
+type Desugared a = Either (Py.ExpressionList a) (Py.Yield a)
+
+-- We have to pair locations and names, and tuple syntax is harder to
+-- read in this case than a happy little constructor.
+data Located a = Located Loc a
+
+-- Desugaring an RHS involves walking as deeply as possible into an
+-- assignment, storing the names we encounter as we go and eventually
+-- returning a terminal expression. We have to keep track of which
+desugar :: (Member (Reader SourcePath) sig, Carrier sig m, MonadFail m)
+        => RHS Span
+        -> m ([Located Name], Desugared Span)
+desugar = \case
+  Left Py.Assignment { left = OneExpression name, right = Just rhs, ann} -> do
+    loc <- locFromTSSpan <$> ask <*> pure ann
+    let cons = (Located loc name :)
+    fmap (first cons) (desugar rhs)
+  Right (Right any) -> pure ([], any)
+  other -> fail ("desugar: couldn't desugar RHS " <> show other)
+
+-- This is a fold function that is invoked from a left fold but that
+-- returns a function (the 'difference' pattern) so that we can pass
+-- information about what RHS we need down the chain: unlike most fold
+-- functions, it has four parameters, not three (since our fold
+-- returns a function). There's some pun to be made on "collapsing
+-- sugar", like "icing" or "sugar water" but I'll leave that as an
+-- exercise to the reader.
+collapseDesugared :: (CoreSyntax syn t, Member (Reader Bindings) sig, Carrier sig m)
+                  => (t Name -> m (t Name)) -- A meta-continuation: it takes a name and returns a continuation
+                  -> Located Name           -- The current LHS to which to assign
+                  -> t Name                 -- The current RHS to which to assign, yielded from an outer continuation
+                  -> m (t Name)             -- The properly-sequenced resolut
+collapseDesugared cont (Located loc n) rem =
+  let assigning = fmap (Core.annAt loc) . fmap ((Name.named' n :<- rem) >>>=)
+  in assigning (local (def n) (cont (pure n))) -- gotta call local here to record this assignment
+
 instance Compile (Py.Assignment Span) where
   compileCC it@Py.Assignment
-    { Py.left = Py.ExpressionList
-      { Py.extraChildren =
-        [ Py.PrimaryExpressionExpression (Py.IdentifierPrimaryExpression (Py.Identifier { Py.bytes = name }))
-        ]
-      }
-    , Py.right = Just rhs
+    { left = OneExpression name
+    , right = Just rhs
+    , ann
     } cc = do
-    value  <- compile rhs
-    let assigning n = (Name.named' name :<- value) >>>= n
-    locate it =<< assigning <$> local (def name) cc
+    p <- ask @SourcePath
+    (names, val) <- desugar rhs
+    let allNames = Located (locFromTSSpan p ann) name : names
+    compile val >>= foldl' collapseDesugared (const cc) allNames >>= locate it
```

Comment:
We could replace this `foldl’` with a `foldr` if we used a `Stack (Located Name)`, but I’m not really sure it matters in the grand scheme of things.

Reply:
1. Using a `Stack` would allow desugaring of assignments to be tail-recursive; probably not a big deal all things considered.

2. Using a left- vs. right-fold does not depend on whehter the bindings are in a list or a stack. It’s the associativity of the operation, not the data.

3. I think this would be a lot simpler as a right fold, specifically it shouldn’t require the difference pattern approach since you can just pass `cc` as the terminal value. Does that work out?

Reply:
@robrix I don’t see a way to make this work without the difference pattern. 

Given `x = y = z = True`, we desugar that to `([“x”, “y”, “z”], True)`. At each stage of the fold, we have to remember what the “last” value is, so that we can put that on the RHS of each assignment (with the initial “last” value being `compile True`). How do we propagate that information without going from the left, and remembering a last value?

As far as `foldr` goes, I think I _do_ need to use a `Stack`, or reverse `allNames` before passing it in. Otherwise I generate `x <- #true; y <- x; z <- y`, and I can’t just reassociate `collapseDesugared` because `>>>=` takes a different type on its left than it does its right.

Am I missing something really obvious?

Reply:
:narrator: He was missing something really obvious.